src/main/java/bio/terra/workspace/db/ResourceDao.java

```diff
@@ -134,6 +141,80 @@ public boolean deleteResource(UUID workspaceId, UUID resourceId) {
         .collect(Collectors.toList());
   }
 
+  /**
+   * @param workspaceId identifier for work space to enumerate
+   * @param controlledResourceIds identifiers of controlled resources visible to the caller
+   * @param resourceType filter by this resource type - optional
+   * @param stewardshipType filtered by this stewardship type - optional
+   * @param offset starting row for result
+   * @param limit maximum number of rows to return
+   * @return list of resources
+   */
+  @Transactional(
+      propagation = Propagation.REQUIRED,
+      isolation = Isolation.SERIALIZABLE,
+      readOnly = true)
+  public List<WsmResource> enumerateResources(
+      UUID workspaceId,
+      List<UUID> controlledResourceIds,
+      @Nullable WsmResourceType resourceType,
+      @Nullable StewardshipType stewardshipType,
+      int offset,
+      int limit) {
+    List<UUID> idList = Optional.ofNullable(controlledResourceIds).orElse(new ArrayList<>());
+
+    MapSqlParameterSource params =
+        new MapSqlParameterSource()
+            .addValue("workspace_id", workspaceId.toString())
+            .addValue("offset", offset)
+            .addValue("limit", limit);
+
+    StringBuilder sb = new StringBuilder(RESOURCE_SELECT_SQL);
+    if (resourceType != null) {
+      sb.append(" AND resource_type = :resource_type");
+      params.addValue("resource_type", resourceType.toSql());
+    }
+
+    // There are three cases for the stewardship type filter
+    // 1. If it is REFERENCED, then we ignore idList and just filter
+    //    for referenced resources.
+    // 2. If it is CONTROLLED, and the idList is not empty, then we filter for
+    //    CONTROLLED and require that the resources be in the idList.
+    // 3. If it is null, then we want the OR of 1 and 2
```

Comment:
This seems to me like we want different public enumerate methods on ResourceDao. When there are some arguments that are only valid with certain values of arguments, I think it's best to define a Java API where invalid argument combinations are less allowed. While our HTTP API can handle all the methods, and do the appropriate request validation on arguments, I don't think we should restrict our Java methods to be 1 to 1 because of that.
Maybe a enumerateControlledResources methods when we want to specify controlledResourceIds.

At least let's explain in the javadoc here and on ResourceService how these parameters interact together.

Reply:
I expanded the javadoc comment to try to clarify. The complexity here is due to a problem I found when passing in the id list. If the list is empty, the resulting SQL is invalid. So instead of simply adding one more filter, like resource type, I end up with this more complex query construction.

I have a bias to invest in the query generation and keep the database work to a single scan/single transaction.